glean-core/android/src/main/java/mozilla/telemetry/glean/scheduler/PingUploadWorker.kt

```diff
@@ -22,6 +22,105 @@ import java.io.FileNotFoundException
+    .build()
+
+/**
+ * Build the [OneTimeWorkRequest] for enqueueing in the [WorkManager].  This also adds a tag
+ * by which enqueued requests can be identified.
+ *
+ * @return [OneTimeWorkRequest] representing the task for the [WorkManager] to enqueue and run
+ */
+internal inline fun <reified W : Worker> buildWorkRequest(tag: String): OneTimeWorkRequest {
+    return OneTimeWorkRequestBuilder<W>()
+        .addTag(tag)
+        .setConstraints(buildConstraints())
+        .build()
+}
+
+/**
+ * This function encapsulates processing of a single ping file.
+ *
+ * @param file The [File] to process
+ *
+ */
+@Suppress("ReturnCount")
+private fun processFile(file: File): Boolean {
+    // This function is from PingsStorageEngine in glean-ac
+
+    var processed = false
+    BufferedReader(FileReader(file)).use {
+        try {
+            val path = it.readLine()
+            val serializedPing = it.readLine()
+
+            processed = serializedPing == null ||
+                Glean.httpClient.doUpload(path, serializedPing, Glean.configuration)
+        } catch (e: FileNotFoundException) {
+            // This shouldn't happen after we queried the directory.
+            Log.e(LOG_TAG, "Could not find ping file ${file.name}")
+            return false
+        } catch (e: IOException) {
+            // Something is not right.
+            Log.e(LOG_TAG, "IO Exception when reading file ${file.name}")
+            return false
+        }
+    }
+
+    return if (processed) {
+        val fileWasDeleted = file.delete()
+        Log.d(LOG_TAG, "${file.name} was deleted: $fileWasDeleted")
+        true
+    } else {
+        // The callback couldn't process this file.
+        false
+    }
+}
+
+/**
+ * Function to deserialize and process all serialized ping files.  This function will ignore
+ * files that don't match the UUID regex and just delete them to prevent files from polluting
+ * the ping storage directory.
+ *
+ * @return Boolean representing the success of the upload task. This may be the value bubbled up
+ *         from the callback, or if there was an error reading the files.
+ */
+fun processDirectory(lock: Any, storageDirectory: File): Boolean {
+    var success = true
+
+    Log.d(LOG_TAG, "Processing persisted pings at ${storageDirectory.absolutePath}")
+
+    synchronized(lock) {
+        storageDirectory.listFiles()?.forEach { file ->
+            if (file.name.matches(Regex(FILE_PATTERN))) {
+                Log.d(LOG_TAG, "Processing ping: ${file.name}")
+                if (!processFile(file)) {
+                    Log.e(LOG_TAG, "Error processing ping file: ${file.name}")
+                    success = false
+                }
+            } else {
+                // Delete files that don't match the UUID FILE_PATTERN regex
+                Log.d(LOG_TAG, "Pattern mismatch. Deleting ${file.name}")
+                file.delete()
+            }
+        }
+    }
+
+    return success
+}
```

Comment:
I'm not _super_ excited about global stuff shared between multiple classes floating around the top of the PingUploadWorker.kt file.  Would this be better to have in a Util file?

Reply:
It's not more or less global than it was before as static functions inside a companion object.

I don't like introducing "Util" files as a catch-all for non-class functions.

Reply:
Okay, not a strong opinion on this, it was just a nit.